src/courtroom.cpp

```diff
@@ -3862,27 +3865,21 @@ void Courtroom::handle_song(QStringList *p_contents)
     }
   }
 
-  int error_code = music_player->play(f_song, channel, looping, effect_flags);
-
-  if (is_stop) {
-    ui_music_name->setText(tr("None"));
-    return;
-  }
-
-  if (error_code == BASS_ERROR_HANDLE) { // Cheap hack to see if file missing
-    ui_music_name->setText(tr("[MISSING] %1").arg(f_song_clear));
-    return;
-  }
-
-  if (f_song.startsWith("http") && channel == 0) {
-    ui_music_name->setText(tr("[STREAM] %1").arg(f_song_clear));
-    return;
+  QFuture<QString> future = QtConcurrent::run(music_player,&AOMusicPlayer::play, f_song, channel, looping, effect_flags);
+  if (channel == 0) {
+    //Current song UI only displays the song playing, not other channels.
+    //Any other music playing is irrelevant.
+    music_player->music_watcher.setFuture(future);
```

Comment:
You can use [`QFuture::cancel`](https://doc.qt.io/qt-5/qfuture.html#cancel) here

Reply:
Unfortunately I already tried that but unless someone gets down and dirty with QThread there's no easy way to cancel.

"Be aware that not all running asynchronous computations can be canceled. For example, the future returned by QtConcurrent::run() cannot be canceled;"